File: 2020/09_EncodingError/cypher.lua
-- ======================================================================
-- Encoding Error
--   Advent of Code 2020 Day 09 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         c y p h e r . l u a
-- ======================================================================
-- A solver for the Advent of Code 2020 Day 09 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Cypher = { part2 = false, text = {}, numbers = {}, preamble = 25 }
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                                 Cypher
-- ======================================================================

-- Object for Encoding Error

function Cypher:Cypher (o)

  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.preable = o.preamble or 25
  o.numbers = {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Cypher:_process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)

  -- 1. Loop for each line of the text
  for _, line in ipairs(text) do

    -- 2. Add the number to the entries
    table.insert(self.numbers, tonumber(line))
    end

end

function Cypher:rule_breaker()
  -- Finds a number that does match the summary rule
  
  -- 1. Set index based on size of preamble
  local test_start = 1 + self.preamble
  local test_end = #self.numbers
  local found = nil
  
  -- 2. Loop for all of the numbers to test
  for tindex = test_start, test_end do
    local test = self.numbers[tindex]
    found = test
    
    -- 3. Loop for all of the previous numbers
    for pindex = tindex - self.preamble, tindex - 2 do
      local previous = self.numbers[pindex]
      
      -- 4. Loop for the remaining previous numbers
      for rindex = pindex + 1, tindex - 1 do
        local remaining = self.numbers[rindex]
        
        -- 5. Check if this pair of previous numbers is equal to the test number
        if previous + remaining == test then
          
          -- 6. This test number is fine, we will need another
          found = nil
        end
      end
    end
    
    -- 7. If this number did not satisify, return it
    if found ~= nil then
      return found
    end  
    
    -- 8. Try another number
  end
  
  -- 9. Every number satisified
  return nil
end

function Cypher:weakness()
  -- Find the weakness in the XMAS code
  
  -- 1. Need the rule breaker from part 1
  local rule_breaker = self:rule_breaker()
  local knt_numbers = #self.numbers
  -- print(rule_breaker, knt_numbers)
  
  -- 2. Loop for all of the numbers (except the last)
  for bindex = 1, knt_numbers -1 do
    local bnumber = self.numbers[bindex]
    local sum = bnumber
    local min = bnumber
    local max = bnumber
      
    -- 3. Loop for the remainin numbers
    for rindex = bindex + 1, knt_numbers do
      local rnumber = self.numbers[rindex]
      sum = sum + rnumber
      min = math.min(min, rnumber)
      max = math.max(max, rnumber)
      -- print(bindex, rindex, sum, min, max, rule_breaker)
        
      -- 4. So these number sum to the rule_breaker?
      if sum == rule_breaker then
          
        -- 5. If so return the minimum and maximum numbers
        return min + max
      end
        
      -- 6. If the sum is greater than the rule_breaker we need a new beginning
      if sum > rule_breaker then
        goto continue
      end
        
    end
    :: continue ::
  end
  
  -- 7. Never found a string of numbers which added up to the rule_breaker
  return nil
end

    
function Cypher:part_one(args)
  -- Returns the solution for part one

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  return self:rule_breaker()
end


function Cypher:part_two(args)
  -- Returns the solution for part two

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part two
  return self:weakness()
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Cypher

-- ======================================================================
-- end                       c y p h e r . l u a                      end
-- ======================================================================

File: 2020/09_EncodingError/cypher.ts
// ======================================================================
// Encoding Error
//   Advent of Code 2020 Day 09 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           c y p h e r . t s
//
// A solver for the Advent of Code 2020 Day 09 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                                 Cypher
// ======================================================================

export class Cypher {
  // Object for Encoding Error
  text: string[];

  part2: boolean;

  numbers: number[];

  preamble: number;

  constructor(text: string[], part2 = false, preamble = 25) {
    // Create a Cypher object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.numbers = [];
    this.preamble = preamble === undefined ? 25 : preamble;

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      for (let indx = 0; indx < this.text.length; indx += 1) {
        this.numbers.push(+this.text[indx]);
      }
    }
  }

  rogue_number(): number {
    // Find the number that doesn't follow the rules

    // 1. Set index based on the preamble
    const testStart = this.preamble;
    const testEnd = this.numbers.length;
    let result = NaN;

    // 2. Loop for all of the number to test
    for (let indx = testStart; indx < testEnd; indx += 1) {
      const test = this.numbers[indx];
      result = test;

      // 3. Loop for the last bunch of number
      for (let pindx = indx - this.preamble; pindx < indx - 1; pindx += 1) {
        const previous = this.numbers[pindx];

        // 4. Loop for the remaining previous numbers
        for (let rindx = pindx + 1; rindx < indx; rindx += 1) {
          const remaining = this.numbers[rindx];

          // 5. Check if this pair of previous number is equal to the test number
          if ((previous + remaining) === test) {
            // 6. The test number is fine, we will need to find another
            result = NaN;
          }
        }
      }
      // 7. If this number is not the result of two of the previous numbers, return it
      if (!Number.isNaN(result)) {
        return result;
      }
      // 8. Well then, try to find another
    }
    // 9. Every number satisified
    return NaN;
  }

  weakness(): number {
    // Find the weakness in the XMAS code, returning min+max

    // 1. Need the rule breaker from part 1
    const ruleBreaker = this.rogue_number();

    // 2. Loop for all the numbers (except the last)
    for (let bindx = 0; bindx < this.numbers.length - 1; bindx += 1) {
      const bottom = this.numbers[bindx];
      let rangeMin = bottom;
      let rangeMax = bottom;
      let rangeSum = bottom;

      // 3. Loop for all of the remaining numbers
      for (let indx = bindx + 1; indx < this.numbers.length; indx += 1) {
        const current = this.numbers[indx];
        rangeSum += current;
        rangeMin = Math.min(rangeMin, current);
        rangeMax = Math.max(rangeMax, current);

        // 4. Have we found the range we were looking for?
        if (rangeSum === ruleBreaker) {
          // 5. If so, return the min plus max of the range
          return rangeMin + rangeMax;
        }

        // 6. If the sum is greater than the rule breaker, we need a new beginning
        if (rangeSum > ruleBreaker) {
          indx = this.numbers.length;
        }
      }
    }
    // 7. Never did find a range that added up to the rule breaker
    return NaN;
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      return this.weakness();
    }
    return this.rogue_number();
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                        c y p h e r . t s                       end
// ======================================================================
